Custom Item Support

KillerOfPie opened the pull request:
- [x] Added command to add custom item from players hand
  * Needs further checks to prevent air and other such items
- [x] Added methods to add custom items
  * Needs remove item method
- [x] Added CustomItem.yml which stores serialized ItemStacks
  * Works for normal items and for potion type but not extended potion data

KillerOfPie committed:
Custom Item Support
- Added command to add custom item from players hand
  * Needs further checks to prevent air and other such items
- Added methods to add custom items
  * Needs remove item method
- Added CustomItem.yml which stores serialized ItemStacks
  * Works for normal items and for potion type but not extended potion data

KillerOfPie commented:
This isn't done yet, but I want you to look it over see if there's anything you want done a different way

KillerOfPie commented:
![screenshot_91](https://user-images.githubusercontent.com/7243280/29747580-ea2481f2-8ab3-11e7-8325-e2a950e0a5c8.png)

Thats what the config looks like with items added and it does look like it works for all potions properly, just messed up when i was testing earlier. Still needs the ability to put it on a sign

This also gets us real close to enchanted books and potions working properly for specific potions

SparklingComet reviewed (approved):
The idea looks good and if we can implement it for enchanted books and potions, we'd squash two bugs with a PR.

I'll apply the ``incomplete`` label for now.

beautifulpoison commented:
Awesome, will keep track here. Thanks for adding it!

SparklingComet commented:
@KillerOfPie any news?

KillerOfPie commented:
@SparklingComet been busy and was having problems with it working on signs but not rn

SparklingComet commented:
@KillerOfPie we could release 1.6.0 now and take more time to develop custom items properly.
We could also wait and release a huge update instead of two big ones... but I do not know whether that is a good idea.

KillerOfPie commented:
I'm having errors rn and I don't know of it was previous stuff or current stuff causing them, so I would wait for 1.6 to have custom items, it's already a pretty big update anyways and I want to see if we can make owner finding any faster(maybe on a separate thread)?

SparklingComet commented:
> I want to see if we can make owner finding any faster(maybe on a separate thread)?

What exactly do you mean by that?

KillerOfPie commented:
If there are too many owners it'll lag a bit, if we throw run that search in a separate thread it may run a little smoother

SparklingComet commented:
> If there are too many owners it'll lag a bit, if we throw run that search in a separate thread it may run a little smoother

Well, you cannot make any Bukkit API-calls inside an async task, which means that you could, indeed, parse the Inventory's title (String#split() etc) in another thread, but you wouldn't be able to end the action without major complications.
I think it would be best if we set a maximum amount of owners+members inside the config, warning administrators not to increase it too much.

KillerOfPie commented:
We do have a max still seems to be a bit slow at times, ill see if i can fix it without another thread tho

SparklingComet commented:
> We do have a max still seems to be a bit slow at times, ill see if i can fix it without another thread tho

if there are too many owners, it means it is too high

KillerOfPie commented:
What

SparklingComet commented:
If it is lagging, aka there are too many owners for the server to process them at a reasonable rate, 

> I think it would be best if we set a maximum amount of owners+members inside the config, warning administrators not to increase it too much.

where the default value is lower than the number of owners you are experiencing

KillerOfPie committed:
Not working idk what was changed

beautifulpoison commented:
Is this still in development?

SparklingComet commented:
Theoretically yes. I am not sure what the progress is, though, as I have not worked on it.
I believe there are a couple bugs that need fixing, but you will need to ask @KillerOfPie about that.

KillerOfPie commented:
@SparklingComet @beautifulpoison I'll take a look at it in the next few days and get a version out for you.

beautifulpoison commented:
Awesome thank you @KillerOfPie

KillerOfPie commented:
@beautifulpoison almost got it done before I left for work, but keep hitting more bugs. Hopefully I have this done some time early next week

SparklingComet commented:
I have just released v1.6.0-beta (refer to #26 ), which does *not* contain the commits in this PR.
To avoid further delays to the release of Multi-Access, we had to postpone Custom items to 1.7.0.

KillerOfPie committed:
- Added support for custom items
- Started support for enchanted books and potions
- changed version to 1.70 BETA
- Code Cleanup

KillerOfPie commented:
@beautifulpoison its ready I tested as much as i could may be some minor bugs but it needs to be in a proper environment for that

KillerOfPie committed:
- Added support for all potions
- Fixed a couple bugs
- Changed isValidType method
- Tweaked item command to show potion names

SparklingComet commented on src/main/java/org/shanerx/tradeshop/Potions.java:
```diff
@@ -0,0 +1,299 @@
+import org.bukkit.potion.PotionType;
+
+@SuppressWarnings("unused")
+public enum Potions {
```
Maybe we want to call this PotionType or something similar, to comply to enum naming conventions (singular nouns).

KillerOfPie replied:
PotionType is already used in bukkit to i wanted to avoid that, and so is Potion. Potions is used in NMS but we will never need to touch that so having this doesnt mean we have 2 enums with the same name

SparklingComet replied:
You can call it TSPotion or something like that.
Not that important though.

KillerOfPie replied:
Im going to leave it for now just because i dont want to go through everything and update the name, maybe in a cleanup later, but when we're doing 2.0 it will definitely get changed

SparklingComet commented on src/main/java/org/shanerx/tradeshop/Potions.java:
```diff
@@ -0,0 +1,299 @@
+        return pot;
+    }
+
+    public static Potions getPotion(String str) {
```
What is the intention behind wrapping the Potions.valueOf() function? Simply renaming the method?

KillerOfPie replied:
isType checks for null and IllegalArguments which prevents errors, this returns the enum vs isType which is only a boolean

SparklingComet commented on src/main/java/org/shanerx/tradeshop/Potions.java:
```diff
@@ -0,0 +1,299 @@
+        }
+
+        try {
+            if (valueOf(type.toUpperCase()) != null) {
```
If the null-check already occurs here, what’s the point in repeating it in getPotion()?

KillerOfPie replied:
getPotion() has no null check, it uses this to check for null and illegal arguments, then reutrns null if nothing is found

SparklingComet commented on src/main/java/org/shanerx/tradeshop/TradeShop.java:
```diff
@@ -250,6 +323,13 @@ private void save() {
                 e.printStackTrace();
             }
+
+        if (CustomItems != null)
```
lowerCamelCase please

KillerOfPie replied:
Blame 2am programming

SparklingComet commented on src/main/java/org/shanerx/tradeshop/Updater.java:
```diff
@@ -27,161 +27,159 @@
+        BEHIND,
+
+        UNKNOWN
+    }
```
Could have been in a separate commit, but fine.

Is this simply a cleanup or am I missing something obvious?

KillerOfPie replied:
I didnt change anything here but i have it reformat the code before i commit, so probably just that

SparklingComet replied:
Makes sense. No objections to that.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/Executor.java:
```diff
@@ -58,29 +60,48 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
                 }
 
-                String[] help = new String[10];
+                StringBuilder sb = new StringBuilder();
```
This is a good one. I don’t know why we made use of arrays, it’s much more efficient this way.

KillerOfPie replied:
we had the array then a loop at the end to add them all to the string builder which makes no sense when you can just use stringbuilder for all of it

SparklingComet commented on src/main/java/org/shanerx/tradeshop/itrade/IShopCreateEventListener.java:
```diff
@@ -54,14 +53,10 @@ public void onSignChange(SignChangeEvent event) {
+            event.setLine(0, "");
+            event.setLine(1, "");
+            event.setLine(2, "");
+            event.setLine(3, "");
```
According to the Spigot docs, Sign#update() should *always* be called. Did it get updated without it?

KillerOfPie replied:
event.setLine updates on its own where s.setline and s.update seem to have problems with actually updating the line

SparklingComet commented on src/main/resources/plugin.yml:
```diff
@@ -1,5 +1,5 @@
 name: TradeShop
-version: 1.6.0-BETA
+version: 1.7.0-BETA
```
The most appropriate tag would be ``-DEV``, but we never used it so...

KillerOfPie replied:
I can start using it, didnt know that you added it

SparklingComet replied:
It’s always been there tbh. It’s purpose is to tell that it’s not supposed to be for public use.

SparklingComet replied:
*production

KillerOfPie replied:
Ill start using it

SparklingComet reviewed (commented):
Overall good work! I didn’t check everything because you said it is working.
Also, splitting relatively large commits into several smaller ones might make it easier to review and debug. Not a huge issue though.
Thank you for an addition we have wanted for so long.

beautifulpoison commented:
I have a live server I'll throw this on to now, and will have another one going live on the 1st that should have a pretty big player count. Thanks for doing this, will keep updated with bugs.

KillerOfPie commented:
@beautifulpoison can you build it yourself or do you need a built jar?

KillerOfPie committed:
- Fixed non-camelCased variables

SparklingComet commented:
Alright, I believe we can merge this into `1.7`.

KillerOfPie commented:
now it should be ready

beautifulpoison commented:
Can I get a jar please

SparklingComet commented:
@beautifulpoison

https://www.spigotmc.org/resources/tradeshop.32762/

beautifulpoison commented:
Thank you

beautifulpoison commented:
With these custom items how do we add them in? I have Crates plugin on my server and want to add keys to be traded for other keys. How would one go about doing this?

beautifulpoison commented:
Hm https://imgur.com/a/3cxsk

SparklingComet commented:
Please share your config/the information about the crate
Also please open an issue, this isn’t really the most appropriate place to discuss it
